glean-core/ffi/src/event.rs

```diff
@@ -0,0 +1,105 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::convert::TryFrom;
+use std::os::raw::c_char;
+
+use ffi_support::{define_handle_map_deleter, ConcurrentHandleMap, FfiStr};
+use lazy_static::lazy_static;
+
+use glean_core::metrics::{EventMetric, MetricType};
+use glean_core::{CommonMetricData, Lifetime};
+
+use crate::handlemap_ext::HandleMapExtension;
+use crate::{
+    from_raw_int_array_and_string_array, from_raw_string_array, RawIntArray, RawStringArray, GLEAN,
+};
+
+lazy_static! {
+    pub static ref EVENT_METRICS: ConcurrentHandleMap<EventMetric> = ConcurrentHandleMap::new();
+}
+define_handle_map_deleter!(EVENT_METRICS, glean_destroy_event_metric);
+
+#[no_mangle]
+pub extern "C" fn glean_new_event_metric(
+    category: FfiStr,
+    name: FfiStr,
+    send_in_pings: RawStringArray,
+    send_in_pings_len: i32,
+    lifetime: i32,
+    disabled: u8,
+    extra_keys: RawStringArray,
+    extra_keys_len: i32,
```

Comment:
@badboy: So here's a use case for the fancy new FFI-generating macro that I couldn't find a solution for.  This is a pair of arguments that need to convert to a single argument in Event::new().  Tips online for this seem to suggest passing the argument as a [struct over FFI](http://jakegoulding.com/rust-ffi-omnibus/tuples/), and then of course we can implement `Into` on that struct, but I couldn't figure out how to create and pass that struct from Kotlin...

Barring that, it looks like `new` is required in the macro, so couldn't just skip that one.

So, I've left this file as not using the macro at all for now (which still works, but lacks the cool factor...)

Reply:
Ah yeah. A single macro will never cover all the cases we have (or become such a big macro that _I_ don't want to maintain it).

The macro should cover the common cases. If we need to fall back to manual implementation here and there that seems fine to me IMO.
(We might still have another/reduced macro that allows skipping the `new` part)

Reply:
`new` can be optional: #174